### ISSUE TITLE
Rethinking internal declaration and loading of config values

### ISSUE THREAD

**Issue by tsipinakis** [dunst-project/dunst#412]
As referenced in #371 having setting parsing and documentation in `settings.c` and splitting default values into `config.h` as global variables is an ugly way to handle setting declaration, I propose the following alternative:

Create a unified `setting` struct that will hold all attributes of a setting, this way all settings can be specified in an array along with possible aliases and ways to handle backwards compatibility.

This might look similar to what already exists in `settings.c` but while it does lose some functionality since we can't add setting-specific code between parsing I would consider that an advantage since it standardizes a way to declare settings without having to go on a case-by-case basis. (Look at all the hacks for backwards compatibility that exist in there!)

Refer to the following data layout:

```c
struct setting {
      char *name; //config key
      char *section; //config section
      char *cmdopt, // cmdline option
      char *description; //description for help menu
      enum setting_type type; // Enum of all setting types - to know what to cast the void* to
      char *default_value;
      void *value;
      void (*parser)(char *cfg_value); // Function pointer for the parser - to be used in case of enums or other special settings
      /* TODO: Think of a way to signify aliases/identical settings 
          for backwards compatibility: i.e. `markup` and `allow_markup`
          maybe also for a way to specify a parser for other deprecated
          settings that affect this one */
}
```

## Storage of values

An idea that I thought about a lot was also migrating the value storage into the same data structure but after consideration I decided to keep the `settings` struct around and simply use a pointer to it in the declaration.

The main argument is that if the storage is done inside the array then each time we want to retrieve the value of a setting a function has to be called which would do a lookup, this has many disadvantages such as extra processing time for the look-up and that there is no convenient way to return a custom type in C so the result would have to be cast from a void pointer each time.

### Default value

Until now default values for settings were set in their proper type inside `config.h` as a global variable (updated to a struct with #371). Both of these are ugly solutions in their own way, ideally both the setting declaration and default value assignment would be together.

The solution that I came up with is to simply store the default value as a string as it would have been set in the config file and have a function call the parser for all settings when dunst starts up.

I've look into (and rejected) the following options

* Void pointer: Can't be easily set to the address of a literal without ugly C hacks
* Union: Possible if we go that route, but has the disadvantage that we have to repeat the type (i.e. `... TYPE_TIME, .time = 1234 ...` and slightly increased memory footprint along with having to add all enums into the union.
* In `settings`: Creates the problem we're trying to solve - setting attributes in different areas). 

## Declaration

With this setup, settings can be declared in an array as so:

```c
struct setting settings[] = {
      {
            "geometry", 
            /*
            Note: can also do it in a slightly longer, more detailed way as follows:
            .name = "geometry"
            */
            "global", 
            "Geometry for the window", 
            TYPE_STRING,  
            "0x0", 
            (void *) &settings.geom, 
            NULL
      },
      {
            "markup",
            "global",
            "Specify how markup should be handled",
            TYPE_ENUM,
            "full",
            (void *) &settings.markup,
            &parse_markup_mode
      },
      .
      .
      .
}
```

**Comment by bebehei** [comment]
> ```
>             /*
>             Note: can also do it in a slightly longer, more detailed way as follows:
>             .name = "geometry"
>             */
> ```

Please prefer this setting in general. I can't even trace back, why there is an additional `NULL` value insert into `geometry` part, but not into `markup`. (While writing this, I've found it now, but 

---

So, how would the processing look like?

```Pseudocode
for (setting in settings) {
    switch(setting->type){
    STRING:
        setting->value = (void*) option_get_string(setting->name, setting->section, setting->cmdlinekey, setting->default, setting->description);
    INT:
        setting->value = (void*) option_get_int(setting->name, setting->section, setting->cmdlinekey, setting->default, setting->description);
    FUNC:
       // call the self-defined function parser here, right?
    default:
          ERROR.
     }
}
```

---

>       /* TODO: Think of a way to signify aliases/identical settings 
>           for backwards compatibility: i.e. `markup` and `allow_markup`
>           maybe also for a way to specify a parser for other deprecated
>           settings that affect this one */

Dunst 2.0?

---

Very cool proposal. Unifying the settings structure is very cool.

I don't like the pointer casts all the time. What about using a `GVariant`? We already include glib and the overhead is minimal, while we avoid special hacks. Also, if you save the default in a GVariant, you can switch based on the `GVariant` type of default (saving the type enum).

**Comment by tsipinakis** [comment]
> I can't even trace back, why there is an additional NULL value insert into geometry part, but not into markup

Probably because I forgot to include it, there was supposed to be a function pointer in there - oops

> So, how would the processing look like?

Yeah, that's basically exactly how I'm thinking of doing the processing

> Dunst 2.0?

There might be more deprecations in the future, so finding a way of handling them now would be wise in my opinion.

> What about using a GVariant?

Interesting idea, will look a bit more into how we can use GVarians here.

> Also, if you save the default in a GVariant, you can switch based on the GVariant type of default (saving the type enum).

To be honest I kind of like having the default as a string as it makes this a bit more readable. `10s` is better than `10*G_USEC_PER_SEC` or `10000000`

**Comment by bebehei** [comment]
- How do we handle `STATIC_CONFIG`? At least with this proposal, static configuration makes absolutely no sense and `config.def.h` and `config.h` has to get removed.

**Comment by tsipinakis** [comment]
> How do we handle STATIC_CONFIG? At least with this proposal, static configuration makes absolutely no sense and config.def.h and config.h has to get removed.

`STATIC_CONFIG` was supposed to disable config parsing so it would make sense even in this scenario but I agree it seems superfluous and will probably deprecate it.

**Comment by bebehei** [comment]
While browsing through the bugtracker, I read that again:

I have to adjust my statements:

> I don't like the pointer casts all the time.  What about using a `GVariant`?

I still don't like it that much, but it's the best to handle it. Better than a `GVariant` ( void pointer is faster, while GVariant is not less error prone).

---
But, there are also basic things still missing:

For example: Backwards compatibility support. Especially, when the logic of the previous setting changes. Or another one: How to handle precedence in different values.

**Comment by bebehei** [comment]
I loved the idea to have a new formalized settings structure and an array of chained single `Setting`, which then get read via code and the corresponding value is put into our settings struct instance. Instead of error prone code, the settings definition becomes rather a typedef.

So I just now found out while writing my previous comment, that the settings pointer of the struct is a hardcoded reference to our settings struct instance. That means, the whole decoupling of settings values and settings query definitions doesn't happen.

So, I started to look at the current `settings.c` file and analyzed it to be able to specify our requirements.

There are only a few patterns, which are used there:
- Simple query of the ini file and directly setting the struct
https://github.com/dunst-project/dunst/blob/48dc3d2458b3dcf8a38171bb701bf0116f2b1362/src/settings.c#L490-L494
- Parameter conversion from string to the needed value
https://github.com/dunst-project/dunst/blob/48dc3d2458b3dcf8a38171bb701bf0116f2b1362/src/settings.c#L181-L201
https://github.com/dunst-project/dunst/blob/48dc3d2458b3dcf8a38171bb701bf0116f2b1362/src/settings.c#L221-L232
https://github.com/dunst-project/dunst/blob/48dc3d2458b3dcf8a38171bb701bf0116f2b1362/src/settings.c#L402-L421
- Precedence handling to use new setting names if set
    (example would be too long)

Do you see, that we use at least three idioms to convert a string to an enum and save it in our settings?

Our settings code looks quite heavy but it's basically a list of code instructions, with a few exceptions. These exceptions let it look heavily like another spaghetti factory. But in reality, it's just a long program, which is not even coupled in most parts. If we minimize these idioms to a single one, we can actually have a long list of uncoupled statements.

wouldn't this suffice the same? And also it would be super type stable.

**Comment by bebehei** [comment]
It makes no sense to put this first into a new major, as we have to test support and have to validate this idea also against backwards compatibility.

I've put this onto the 1.4 milestone. My glassbowl tells me, that 1.4 is the last release before a new major. So it might be the last release.

**Comment by bebehei** [comment]
```c
struct setting {
      char *ini; // config configuration tree
      char *cmdopt; // command line option
      char *description; //description for help menu
```

I would use a single string for ini and a single string for cmdopts. While `ini`'s format is `section:option/fallbacksection:fallbackoption/...` and `cmdopt`'s format is `-cmdopt/-cmdopt/...`

This makes it more easy to rename options in the ini file and hold up backwards compatibility.

---

@tsipinakis You accidentally forgot in the initial struct the `cmdopt` field. I've added it in your initial comment.

**Comment by bebehei** [comment]
Considering the `-help` output, I'd like to have this improved, too. IMHO this is currently a mess and should get improved/sorted in some way.

**Comment by bebehei** [comment]
Dunst currently ignores invalid configuration options and command line flags. So there is no actual guarantee, that the flag you specified actually works.

Dunst assembles the settings this way: There's field XY, which is either in section A and saved under key B or it's available on the commandline with flag C. Search for it and if found, convert the string to its configuration representation. If one of these steps fails, take the default and put this into the settings. So it traverses all fields from the settings struct and searches for every field the given ini and cmdline options.

All options set in dunst are now the arbitary cut of possible options (settings struct) and the set of configured options (ini file).

But I wouldn't go this way. I would make the options read from the inifile/cmdline a subset of all possible options. So that you traverse the commandline and inifile and then search how you can process the specific flag.

This allows dunst to recognize trash, typos etc. and will give a tighter user feedback.

**Comment by fwSmit** [comment]
Implemented by #803
